service_change.go

```diff
@@ -30,15 +37,62 @@ func (w *ServiceChangeWaiter) RefreshFunc() resource.StateRefreshFunc {
 			return nil, "", err
 		}
 
-		return service, service.State, nil
+		// If this is a Kafka service, wait for its component parts (i.e. Kafka Connect,
+		// Kafka REST, and Schema Registry) to be ready
+		state := service.State
+		if state == aivenTargetState && service.Type == "kafka" {
+			if !kafkaServicesReady(service) {
+				state = aivenKafkaServicesStartingState
+			}
+		}
+
+		return service, state, nil
+	}
+}
+
+// If any of Kafka Rest, Schema Registry, and Kafka Connect are enabled, refresh
+// their state to check if they're ready
+func kafkaServicesReady(service *aiven.Service) bool {
+	kafkaRestEnabled, schemaRegistryEnabled, kafkaConnectEnabled := false, false, false
+	userConfig := service.UserConfig.(map[string]interface{})
+	if enabled, ok := userConfig["kafka_rest"]; ok {
+		kafkaRestEnabled = enabled.(bool)
+	}
+	if enabled, ok := userConfig["schema_registry"]; ok {
+		schemaRegistryEnabled = enabled.(bool)
+	}
+	if enabled, ok := userConfig["kafka_connect"]; ok {
+		kafkaConnectEnabled = enabled.(bool)
 	}
+
+	if !kafkaRestEnabled && !schemaRegistryEnabled && !kafkaConnectEnabled {
+		return true
+	}
+
+	if kafkaRestEnabled {
+		if _, err := http.Get(service.ConnectionInfo.KafkaRestURI); err != nil {
```

Comment:
Not sure what Go will do with that `_`. From the doc: `Caller should close resp.Body when done reading from it.` Seems like assigning that to `_` will make the `rest.Body` leak.

https://golang.org/pkg/net/http/#Get

Reply:
The connection will already have been made if there is no error and will thus be open, regardless of us ignoring it or not. Therefor, one should still close the body.

A simple `rsp.Body.Close()` to signal that the connection can be closed after the error check will do.